src/app/CommandSender.h

```diff
@@ -53,96 +53,158 @@ namespace app {
+     * 1. Path-specific errors go to OnResponse instead of OnError
+     *       - Note: Non-path-specific errors still go to OnError.
+     * 2. Instead of having new parameters at the end of the arguments list, with defaults,
+     *    as functionality expands, a parameter whose type is defined in this header is used
+     *    as the argument to the callbacks
+     *
+     * To support batch commands client needs to be using ExtendedCallback
+     */
+    class ExtendedCallback
+    {
+    public:
+        virtual ~ExtendedCallback() = default;
 
         /**
-         * OnResponse will be called when a successful response from server has been received and processed. Specifically:
+         * OnResponse will be called for all path specific responses from the server that has been received
+         * and processed. Specifically:
          *  - When a status code is received and it is IM::Success, aData will be nullptr.
+         *  - When a status code is received and it is IM and/or cluster error, aData will be nullptr.
+         *      - These kinds of errors are referred to as path-specific errors.
          *  - When a data response is received, aData will point to a valid TLVReader initialized to point at the struct container
          *    that contains the data payload (callee will still need to open and process the container).
          *
          * The CommandSender object MUST continue to exist after this call is completed. The application shall wait until it
          * receives an OnDone call to destroy the object.
          *
-         * It is advised that subclasses should only override this or `OnResponseWithAdditionalData`. But, it shouldn't actually
-         * matter if both are overridden, just that `OnResponse` will never be called by CommandSender directly.
-         *
          * @param[in] apCommandSender The command sender object that initiated the command transaction.
          * @param[in] aPath           The command path field in invoke command response.
          * @param[in] aStatusIB       It will always have a success status. If apData is null, it can be any success status,
          *                            including possibly a cluster-specific one. If apData is not null it aStatusIB will always
          *                            be a generic SUCCESS status with no-cluster specific information.
          * @param[in] apData          The command data, will be nullptr if the server returns a StatusIB.
+         * @param[in] aAdditionalResponseData
+         *                            Additional response data that comes within the InvokeResponseMessage.
          */
         virtual void OnResponse(CommandSender * apCommandSender, const ConcreteCommandPath & aPath, const StatusIB & aStatusIB,
-                                TLV::TLVReader * apData)
+                                TLV::TLVReader * apData, const AdditionalResponseData & aAdditionalResponseData)
```

Comment:
So here is a question: Why not just have a single `ResponseData` argument that encapsulates things like the status and TLVReader and whatnot, if we are doing this green-field anyway?

Reply:
That's a good idea. Since many iterations got to this incremental version, I think it's a good call to bring the response data all in the one struct.